src/wasm/mod.rs

```diff
@@ -38,61 +47,91 @@ impl Server {
 
-    pub fn process(&mut self, msg: String) -> ServerResponse {
-        let mut lines = msg.lines();
+    pub fn register_buckets_callback(&mut self, f: Function) {
+        self.callbacks.register_buckets_callback(f);
+    }
 
-        if lines.clone().count() > 2 {
-            // Skip content length and spacer
-            lines.next();
-            lines.next();
+    pub fn get_buckets(&mut self) -> Promise {
+        let callbacks = self.callbacks.clone();
 
-            let mut content = String::new();
+        future_to_promise(async move {
+            let result = callbacks.get_buckets().await;
 
-            for line in lines {
-                content = content.add(line);
+            match result {
+                Ok(v) => Ok(JsValue::from(v.join(","))),
+                Err(e) => Err(JsValue::from(e)),
             }
+        })
+    }
+
+    pub fn process(&mut self, msg: String) -> Promise {
+        let handler = self.handler.clone();
+        let callbacks = self.callbacks.clone();
+
+        future_to_promise(async move {
+            let lines = msg.lines();
+
+            if lines.clone().count() > 2 {
+                let content = lines
+                    .skip(2)
+                    .fold(String::new(), |c, l| c.add(l));
 
-            if let Ok(req) = utils::parse_request(content) {
-                match self.handler.handle(req) {
-                    Ok(response) => {
-                        if let Some(response) = response {
-                            return ServerResponse {
-                                message: Some(utils::wrap_message(
-                                    response,
-                                )),
-                                error: None,
-                            };
-                        } else {
-                            return ServerResponse {
-                                message: None,
-                                error: None,
-                            };
+                if let Ok(req) =
+                    utils::create_polymorphic_request(content)
+                {
+                    let ctx = RequestContext::new(callbacks.clone());
+                    let mut h = handler.borrow_mut();
+                    match (*h).handle(req, ctx).await {
+                        Ok(response) => {
+                            if let Some(response) = response {
+                                return Ok(JsValue::from(
+                                    ServerResponse {
+                                        message: Some(
+                                            utils::wrap_message(
+                                                response,
+                                            ),
+                                        ),
+                                        error: None,
+                                    },
+                                ));
+                            } else {
+                                return Ok(JsValue::from(
```

Comment:
curiosity: When/how does a response go through `Some(response) = response` and fall into this block here?

Reply:
some requests don't require a response (didChange is one) the request updates the server state then the handler returns None to say, nothing to reply with